components/logins/src/encryption.rs

```diff
@@ -83,23 +83,20 @@ impl EncryptorDecryptor {
 //     - If it returns true, then it's safe to assume the key can decrypt the DB data
 //     - If it returns false, then the key is no longer valid.  It should be regenerated and the DB
 //       data should be wiped since we can no longer read it properly
-pub fn create_canary(text: &str, key: &str) -> ApiResult<String> {
-    handle_error! {
-        EncryptorDecryptor::new(key)?.encrypt(text)
-    }
+#[handle_error(LoginsStorageError)]
+pub fn create_canary(text: &str, key: &str) -> Result<String> {
+    EncryptorDecryptor::new(key)?.encrypt(text)
 }
 
-pub fn check_canary(canary: &str, text: &str, key: &str) -> ApiResult<bool> {
-    handle_error! {
-        Ok(EncryptorDecryptor::new(key)?.decrypt(canary)? == text)
-    }
+#[handle_error(LoginsStorageError)]
+pub fn check_canary(canary: &str, text: &str, key: &str) -> Result<bool> {
+    Ok(EncryptorDecryptor::new(key)?.decrypt(canary)? == text)
 }
 
-pub fn create_key() -> ApiResult<String> {
-    handle_error! {
-        let key = jwcrypto::Jwk::new_direct_key(None)?;
-        Ok(serde_json::to_string(&key)?)
-    }
+#[handle_error(LoginsStorageError)]
+pub fn create_key() -> Result<String> {
+    let key = jwcrypto::Jwk::new_direct_key(None)?;
+    Ok(serde_json::to_string(&key)?)
```

Comment:
I think this is overall a lot cleaner, but my one concern is that it's silently changing the result type.  After `handle_error` is done doing it's wrapping, this will return a `ApiResult<String>`, but you wouldn't think that based on the signature.

I'm not sure this is a great solution, but what if we reversed things?  Then this code would be:
```
#[handle_error(LoginsError)]
pub fn create_key() -> ApiResult<String> {
    let key = jwcrypto::Jwk::new_direct_key(None)?;
    Ok(serde_json::to_string(&key)?)
}
```

That seems a bit more clear to me, although still slightly magical.

Reply:
Yeah there is a balance here of where to put the magic
1. In the one case, the function you write returns `ApiResult` but you wrote it to return `Result` and the signature says `Result`
2. in the other case, the function returns `ApiResult` (and so does the signature) but you have to return `Result` in its body

neither is perfect and I'm happy to change it to be the reverse!

What do you think about changing the name to something like `#[expose_error_as(LoginsStorageError)]`

That way we keep option 1 (i.e the way it is in this PR) and the magic is slightly alleviated because the macro tells us that we are mapping the error

```rs
#[expose_error_as(LoginsStorageError)]
pub fn create_key() -> Result<String> {
    let key = jwcrypto::Jwk::new_direct_key(None)?;
    Ok(serde_json::to_string(&key)?)
}
```

Reply:
I think 1 could work with a nice name, but I'm still not sure what that is.  One issue for me is that the macro is doing several things: coverting, logging, and reporting errors.  The other thing is that by converting the errors, you're changing the `Result<>` type you're returning.  "expose error" doesn't communicate that to me, but I'm struggling to suggest something better. 

 What if we went more verbose?  Something like `#[handle_error_and_convert_to_api_result(LoginsError)]`.